htmresearch/frameworks/nlp/classify_keywords.py

```diff
@@ -130,3 +131,57 @@ def _densifyPattern(self, bitmap, n):
     for i in bitmap:
       sparsePattern[i] = 1.0
     return sparsePattern
+
+
+  def _inferDocumentDetailed(self, tokenList, sortResults=True):
+    """
+    Run inference on the model with this list of tokens and return classification
+    results, sampleIds and distances.  By default this routine will tokenize the
+    document and classify using these tokens. A reset is issued after inference.
+    Repeated sampleIds ARE removed from the results.
+
+    @param tokenList (str)     The list of tokens for inference
+    @param sortResults (bool) If true the list of sampleIds and distances
+                              will be sorted in order of increasing distances.
+
+    @return  (numpy array) An array of size numLabels. Position i contains
+                           the likelihood that this token belongs to the i'th
+                           category. An array containing all zeros implies no
+                           decision could be made.
+             (list)        A list of unique sampleIds
+             (numpy array) An array of distances from this document to each
+                           sampleId
+    """
+    # This model calls for exact matching encodings, which is handled
+    # implicitly by the kNN classifier for category votes, but not for the
+    # returned distances -- i.e. the distances to all prototypes are returned,
+    # whereas a category will only be predicted if there is an exact matching
+    # prototype.
+
+    # For each token run inference on the token and accumulate sum of distances
+    # from this token to all other sampleIds.
+    lastTokenIndex = len(tokenList) - 1
+    categoryVotes = numpy.zeros(self.numLabels)
+
+    for i, token in enumerate(tokenList):
+      votes, idList, distances = self.inferToken(token,
+                                                 reset=int(i == lastTokenIndex),
+                                                 returnDetailedResults=True,
+                                                 sortResults=False)
+
+      if votes.sum() == 0:
+        # No exact matches
+        infDistances = numpy.empty(distances.shape)
+        infDistances[:] = numpy.inf
+        return categoryVotes, idList, infDistances
+
+      # Increment all because a vote implies an exact match
+      categoryVotes[numpy.where(votes > 0)[0]] = 1
+
+      # We only care about 0 distances (exact matches), and disregard all others
```

Comment:
What happens if sortResults = False?  (I'm actually tempted to remove that option from the method signatures - what do you think?)